libs/api/bpf_syscall.cpp

```diff
@@ -6,188 +6,230 @@
 #include <windows.h>
 #include <WinError.h>
 #include <linux/bpf.h>
+#include <sal.h>
+#include <stdexcept>
 
-#define CHECK_SIZE(last_field_name)                                                    \
-    if (!tail_is_zero(                                                                 \
-            attr,                                                                      \
-            offsetof(union bpf_attr, last_field_name) + sizeof(attr->last_field_name), \
-            sizeof(union bpf_attr))) {                                                 \
-        return -EINVAL;                                                                \
-    }
-
-static bool
-tail_is_zero(const void* buf, unsigned int start, unsigned int end)
+template <typename T> class ExtensibleStruct
 {
-    const unsigned char* p = (const unsigned char*)buf + start;
-    const unsigned char* e = (const unsigned char*)buf + end;
+  private:
+    void* _source;
+    size_t _copy_size;
+    T _tmp;
+    T* _ptr;
 
-    for (; p < e; p++) {
-        if ((*p) != 0) {
-            return false;
+    static void
+    check_tail(_In_reads_bytes_(end) const void* buf, size_t start, size_t end)
+    {
+        const unsigned char* p = (const unsigned char*)buf + start;
+        const unsigned char* e = (const unsigned char*)buf + end;
+
+        for (; p < e; p++) {
+            if ((*p) != 0) {
+                throw std::runtime_error("Non-zero tail");
+            }
         }
     }
 
-    return true;
-}
+  public:
+    ExtensibleStruct(_In_reads_bytes_(ptr_size) void* ptr, size_t ptr_size) : _source(ptr)
+    {
+        if (ptr_size >= sizeof(T)) {
+            // Forward compatibility: allow a larger input as long as the
+            // unknown fields are all zero.
+            check_tail(ptr, sizeof(T), ptr_size);
+            _copy_size = 0;
+            _ptr = (T*)ptr;
+        } else {
+            // Backwards compatibility: allow a smaller input by implicitly zeroing all
+            // missing fields.
+            memcpy(&_tmp, ptr, ptr_size);
+            _copy_size = ptr_size;
+            _ptr = &_tmp;
+        }
+    }
+
+    ~ExtensibleStruct() { memcpy(_source, &_tmp, _copy_size); }
+
+    T*
+    operator->()
+    {
+        return _ptr;
+    }
+
+    T*
+    operator&()
+    {
+        return _ptr;
+    }
+
+    T
+    operator*()
+    {
+        return *_ptr;
+    }
+};
 
 int
-bpf(int cmd, union bpf_attr* attr, unsigned int size)
+bpf(int cmd, union bpf_attr* p, unsigned int size)
```

Comment:
"p" is an abbreviation.  "attr" was ok because the term matched "bpf_attr" which was defined outside this repo.

Reply:
Changed this back to attr and made the other uses of attr more specific.